xtask/src/bench.rs

```diff
@@ -1,66 +1,33 @@
-use crate::{
-    cli::{bench::Params, Backend},
-    timer::Timer,
-    Action,
-};
+use crate::{backend::Backend, cli::bench::Params, timer::Timer, workload::Workload};
 use anyhow::Result;
 
-fn rand_keys(len: usize) -> Vec<[u8; 32]> {
-    // keys must be uniformly distributed, but we don't want to spend time on a good hash. So
-    // the next best option is to use a seeded PRNG.
-    use rand::{RngCore as _, SeedableRng as _};
-    let mut keys: Vec<[u8; 32]> = vec![];
-    for i in 0..len {
-        let mut seed = [0; 16];
-        seed[0..8].copy_from_slice(&i.to_le_bytes());
-        let mut rng = rand_pcg::Lcg64Xsh32::from_seed(seed);
-        let mut key = [0u8; 32];
-        for i in 0..8 {
-            let rand = rng.next_u32().to_be_bytes();
-            key[i] = rand[0];
-            key[i + 1] = rand[1];
-            key[i + 2] = rand[2];
-            key[i + 3] = rand[3];
-        }
-        keys.push(key);
-    }
-    keys
-}
-
-pub fn bench(mut params: Params) -> Result<()> {
-    let actions = params.workload.get_actions(params.workload_size);
+pub fn bench(params: Params) -> Result<()> {
+    let workload = Workload::new(
+        params.workload.name.as_str(),
+        params.workload.size,
+        params
+            .workload
+            .initial_capacity
+            .and_then(|s| Some((s as u64) << 2))
```

Comment:
Nothing serious, but doesn't << lead to a panic on overflow?

Reply:
Yes, it could panic. I didn't think about it. Being inside the construction of benchmarking, I didn't put enough attention to panic!
It can be solved enforcing a bound with clap.
But the biggest problem here is that the shift is completely wrong... It should be (1 << s) to be like 2^s. I will correct it now!